### DISCUSSION TITLE
ADNLPModels

### DISCUSSION THREAD

**Discussion by PierreMartinon** [control-toolbox/CTDirect.jl#111] [General]
@ocots @jbcaillau @gergaud @tmigot @amontoison @gdalle
The [issue with Symbolics](https://github.com/control-toolbox/CTDirect.jl/issues/88) has been resolved, however the ADNLPModel topic is still relevant, so I propose we continue it here !

**Comment by PierreMartinon** [comment]
> The next step is to use DifferentiationInterface.jl. We could have for a speed-up for performing the AD because I expect the tapes to be better allocated.
> 
> But the main speed-up that I expect will be related to the use of the symmetry for the coloring of the Hessian.
> 
> ![Messenger_creation_8483570861670037](https://private-user-images.githubusercontent.com/35051714/339847581-8d472499-eda4-4b37-b268-766a1e9483aa.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg0MzQ2MTUsIm5iZiI6MTcxODQzNDMxNSwicGF0aCI6Ii8zNTA1MTcxNC8zMzk4NDc1ODEtOGQ0NzI0OTktZWRhNC00YjM3LWIyNjgtNzY2YTFlOTQ4M2FhLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxNVQwNjUxNTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMmQ5MzQ0MmQwZDFlMDY0ZTIyYTlhYTAzZjJiOTcwMTkwNGRkMjZiZThkMjU0YWM2NmYzNjY2MDEzNmExNmU0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.AaJdnmGkITUgl-XJ-cCuTQnr5pcR9Fh_1QzATVCsjug) 
> If we have less colors, it means less directional derivates. We only use column coloring in `ADNLPModels.jl` for Hessian and Jacobian right now. I can add it for a release 0.8.2.

This certainly looks promising ! 
With the old bocop2 I remember that the computation time was roughly divided 50-50 (well, let's say from 1/3-2/3 to 2/3-1/3) between the linear solver part and the NLP problem functions, including derivatives. Among derivatives the Hessian was typically > 90-95% of the total time, apart from a handful of problems with difficult convergence where the Jacobian actually took a significant part.
So I would expect improvements of the Hessian to have the most impact on CTDirect. I'll have to check whether NLPModelIpopt allows acces to the detailed computational times.

**Reply by tmigot** [reply]
Hi @PierreMartinon !
How do you usually measure the time spend in the evaluations of the model?
I was thinking that it might be interesting to have a BenchmarkNLPModel model modifiers that also keep some statistics on the evaluations. The usage would be something like this:
```
nlp = ADNLPModel("blah blah")
bench_nlp = BenchmarkNLPModel(nlp)
ipopt(bench_nlp)
nlp.counters # would return counters on the number of evaluations as usual
nlp.statistics # would also return basic time/memory/allocations statistics
```
The idea being that BenchmarkNLPModel is only used to improve the implementation as I am suspecting there is an overhead in keeping such statistics that might not be useful for general purpose (or maybe not).

**Reply by gdalle** [reply]
That would be interesting, and probably easy to do with https://github.com/KristofferC/TimerOutputs.jl

**Reply by PierreMartinon** [reply]
Hi everyone
We can use the *print_timing_statistics* option for ipopt, here on a discretized problem to have a better access to the solution.
```
docp = directTranscription(ocp, grid_size=200, init=(state=[1,0.2,0.5], control=0.5))
dsol = solve(docp, print_level=5, tol=1e-12, print_timing_statistics="yes")
println("\n Discrete solution")
println(dsol)
```
This will give the following output (scroll to bottom)
```
 ...
  76 -1.0125751e+00 2.22e-16 7.57e-15 -12.3 3.02e-09    -  1.00e+00 1.00e+00h  1

Number of Iterations....: 76

                                   (scaled)                 (unscaled)
Objective...............:  -1.0125751272882346e+00   -1.0125751272882346e+00
Dual infeasibility......:   7.5749178896056918e-15    7.5749178896056918e-15
Constraint violation....:   2.2204460492503131e-16    2.2204460492503131e-16
Variable bound violation:   9.0405446010308409e-09    9.0405446010308409e-09
Complementarity.........:   5.0004402930230759e-13    5.0004402930230759e-13
Overall NLP error.......:   5.0004402930230759e-13    5.0004402930230759e-13


Number of objective function evaluations             = 81
Number of objective gradient evaluations             = 77
Number of equality constraint evaluations            = 81
Number of inequality constraint evaluations          = 81
Number of equality constraint Jacobian evaluations   = 78
Number of inequality constraint Jacobian evaluations = 78
Number of Lagrangian Hessian evaluations             = 76
Total seconds in IPOPT (w/o function evaluations)    =      0.324
Total seconds in NLP function evaluations            =     22.951


Timing Statistics:

OverallAlgorithm....................:     22.719 (sys:      0.552 wall:     23.275)
 PrintProblemStatistics.............:      0.000 (sys:      0.000 wall:      0.000)
 InitializeIterates.................:      0.025 (sys:      0.000 wall:      0.025)
 UpdateHessian......................:     21.373 (sys:      0.523 wall:     21.899)
 OutputIteration....................:      0.003 (sys:      0.000 wall:      0.003)
 UpdateBarrierParameter.............:      0.248 (sys:      0.003 wall:      0.251)
 ComputeSearchDirection.............:      0.032 (sys:      0.000 wall:      0.032)
 ComputeAcceptableTrialPoint........:      0.190 (sys:      0.002 wall:      0.191)
 AcceptTrialPoint...................:      0.002 (sys:      0.000 wall:      0.002)
 CheckConvergence...................:      0.848 (sys:      0.024 wall:      0.872)
PDSystemSolverTotal.................:      0.260 (sys:      0.003 wall:      0.263)
 PDSystemSolverSolveOnce............:      0.254 (sys:      0.003 wall:      0.256)
 ComputeResiduals...................:      0.004 (sys:      0.000 wall:      0.004)
 StdAugSystemSolverMultiSolve.......:      0.253 (sys:      0.002 wall:      0.256)
 LinearSystemScaling................:      0.000 (sys:      0.000 wall:      0.000)
 LinearSystemSymbolicFactorization..:      0.002 (sys:      0.000 wall:      0.002)
 LinearSystemFactorization..........:      0.000 (sys:      0.000 wall:      0.000)
 LinearSystemBackSolve..............:      0.068 (sys:      0.001 wall:      0.069)
 LinearSystemStructureConverter.....:      0.000 (sys:      0.000 wall:      0.000)
  LinearSystemStructureConverterInit:      0.000 (sys:      0.000 wall:      0.000)
QualityFunctionSearch...............:      0.011 (sys:      0.000 wall:      0.011)
TryCorrector........................:      0.000 (sys:      0.000 wall:      0.000)
Task1...............................:      0.003 (sys:      0.000 wall:      0.003)
Task2...............................:      0.003 (sys:      0.000 wall:      0.003)
Task3...............................:      0.001 (sys:      0.000 wall:      0.001)
Task4...............................:      0.000 (sys:      0.000 wall:      0.000)
Task5...............................:      0.001 (sys:      0.000 wall:      0.001)
Task6...............................:      0.000 (sys:      0.000 wall:      0.000)
Function Evaluations................:     22.398 (sys:      0.549 wall:     22.951)
 Objective function.................:      0.001 (sys:      0.000 wall:      0.001)
 Objective function gradient........:      0.001 (sys:      0.000 wall:      0.001)
 Equality constraints...............:      0.136 (sys:      0.002 wall:      0.138)
 Inequality constraints.............:      0.002 (sys:      0.000 wall:      0.002)
 Equality constraint Jacobian.......:      0.010 (sys:      0.000 wall:      0.010)
 Inequality constraint Jacobian.....:      0.876 (sys:      0.024 wall:      0.901)
 Lagrangian Hessian.................:     21.371 (sys:      0.523 wall:     21.898)

EXIT: Optimal Solution Found.


```
Unfortunately these additional statistics do not seem to appear in the returned solution. Maybe we could redirect and parse the output from ipopt ? Iirc there is an option to save the output as a text file (default result.out). Or we could ask the folks at NLPModelIpopt to retrieve more data :p
```
 Discrete solution
Generic Execution stats
  status: first-order stationary
  objective value: -1.0125751272882346
  primal feasibility: 2.220446049250313e-16
  dual feasibility: 7.574917889605692e-15
  solution: [1.0  2.1911467851546995e-38  1.0  1.00000128100345 ⋯ 0.2020026931250076]
  multipliers: [3.9457885819737464  0.14842616950746015  0.05398054299536937  7.152298916917322e-5 ⋯ 5.102583612261047e-14]
  multipliers_L: [4.999999974876204e-5  4.9999999999999996e-5  0.0  3.8729563425161217e-7 ⋯ 2.6041300037034578e-12]
  multipliers_U: [2.4999998500000082e-12  0.0  0.0  2.5000158626437724e-12 ⋯ 0.0]
  iterations: 76
  elapsed time: 23.275000000000002
  solver specific:
    real_time: 23.275540113449097
    internal_msg: :Solve_Succeeded
```

As for the overhead, the [doc](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_timing_statistics) says "The overall algorithm time is unaffected by this option.", which seems a bit weird.

**Reply by jbcaillau** [reply]
wasn't aware of the `print_timing_statistics`option: nice 🙂!

**Reply by tmigot** [reply]
@PierreMartinon wasn't aware of this too. I made an issue https://github.com/JuliaSmoothOptimizers/NLPModelsIpopt.jl/issues/119 for this

**Reply by amontoison** [reply]
@jbcaillau @ocots @PierreMartinon
New benchmarks on the optimal power flow problems are available on Julia discourse:
https://discourse.julialang.org/t/ac-optimal-power-flow-in-various-nonlinear-optimization-frameworks/78486/94?u=amontoison

The good news is that we still don't exploit the symmetry of the Hessian (so more directional derivates to perform) and more performance improvements from `SparseConnectivityTracer.jl` are coming.

**Reply by tmigot** [reply]
Hey guys! A small follow-up if you want to track timer in NLPModels, I made a TimerNLPModels https://github.com/tmigot/TimerNLPModels.jl . Let me know if you find this useful :).
```julia
using NLPModelsIpopt, ADNLPModels, TimerNLPModels
nlp = TimerNLPModel(ADNLPModel(x -> sum(x.^2), ones(3)))
stats = ipopt(nlp, print_level = 0)
get_timer(nlp)
```
would return
```
 ────────────────────────────────────────────────────────────────────────────
                                    Time                    Allocations
                           ───────────────────────   ────────────────────────
     Tot / % measured:          64.1s /   2.4%            296MiB /  56.0%

 Section           ncalls     time    %tot     avg     alloc    %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 grad!                  3    751ms   47.9%   250ms   58.5MiB   35.3%  19.5MiB
 hess_coord!            1    584ms   37.2%   584ms   97.7MiB   58.9%  97.7MiB
 obj                    2    145ms    9.3%  72.6ms   8.33MiB    5.0%  4.16MiB
 hess_structure!        1   88.1ms    5.6%  88.1ms   1.34MiB    0.8%  1.34MiB
 ────────────────────────────────────────────────────────────────────────────

```

**Reply by jbcaillau** [reply]
@tmigot cute!

**Reply by ocots** [reply]
@0yassine0 will try it here: https://github.com/0Yassine0/COTS.jl

**Reply by ocots** [reply]
@PierreMartinon @jbcaillau how do we handle this? Update of `CTDirect.jl` or let the user do it?

**Reply by PierreMartinon** [reply]
Hmm, the typical user may not use these timing statistics, so keep it on a case-by-case basis ?

**Reply by jbcaillau** [reply]
@ocots @PierreMartinon let's do it the easy way: as we also offer the possibility to retrieve the model (instead of merely `solve`ing), the following is OK (and, yes, in the user's hands):
```julia
using OptimalControl
using NLPModelsIpopt
using TimerNLPModels

@def ocp ...
nlp = direct_transcription(ocp; model=:adnlp) # Return an ADNLPModel
nlp = TimerNLPModel(nlp)
stats = ipopt(nlp; print_level = 0)
get_timer(nlp)
```

**Comment by ocots** [comment]
Ping @0Yassine0.

**Comment by amontoison** [comment]
:rocket: https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/issues/258 :rocket:

**Comment by amontoison** [comment]
@PierreMartinon @jbcaillau @ocots @0Yassine0 @gergaud
I just released ADNLPModels.jl v0.8.3. :metal: 

We use a symmetric coloring now instead of a column coloring for the Hessian of the Lagrangian.
This leads to fewer directional derivatives to compute all nonzeros of the Hessian because we exploit the symmetry.

If you use an optimization solver that needs the Hessian of the Lagrangian (like Ipopt, KNITRO or MadNLP), you will (probably) notice a significant speed-up. In the latest benchmarks, we observed a speed-up greater than x100 :rocket: on some problems:  
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/271#issuecomment-2237489689

These problems likely have a Hessian with an arrowhead form, and exploiting symmetry is particularly important in such cases.
It could highly impact you. :wink: 

Thanks @gdalle for your implementation of the star coloring :+1: 
It could be also relevant for you @frapac.

**Comment by PierreMartinon** [comment]
Hi everyone,
Been working on `CTDirect`, I did a crude benchmark with 8 problems. Times for increasing grid sizes are below for ADNLPModels 0.8.2 and 0.8.3 (in seconds, excluding precompilation, which takes 50-55s). These runs are for MUMPS with MKL libs. Note that the scaling wrt grid size is not perfect since the number of iterations does change for different grid sizes (eg several problems converged with fewer iterations for N=500 vs N=400, leading to very close overall times).

| N     | 100 | 200 | 300 | 400 | 500 | 1000 | 5000 |
|-------|---|----|-----|-----|-----|-----|------|
| 0.8.2 | 9 | 27 | 115 | 340 | 362 | ? | ? |
| 0.8.3 | 6 | 9  | 14  | 45  | 46  | 268 | 2300 |
| speedup | 1.5 | 3.0 | 8.2 | 7.6 | 7.9 | ? | ? |

Not bad at all :D !

**Reply by jbcaillau** [reply]
thank you @amontoison @gdalle and @PierreMartinon 👍🏽

**Reply by gdalle** [reply]
With ADNLPModels v0.8.4 (just released), a few improvements to sparsity detection will land right on your doorstep (including reduced compilation time and some array overloads like `norm`)

**Reply by jbcaillau** [reply]
well well well. giving it a try 🤞🏾

**Comment by amontoison** [comment]
@PierreMartinon @jbcaillau @ocots @0Yassine0 @gergaud
I just released ADNLPModels.jl v0.8.7. :metal: :metal: 

It provides the opportunity to provide the sparsity pattern of the Jacobians and Hessians
directly if it we already know from the applications or if it was already computed before and dumped in a file.
Tangi did a tutorial about it: https://jso.dev/ADNLPModels.jl/dev/sparsity_pattern/

I also switched to the acycling coloring that I implemented in SparseMatrixColorings.jl for sparse Hessian,
it leads to fewer colors than star or column colorings and thus less directional derivates to recover all non-zeros of the Hessian.
The last benchmark shows a speed-up between 10% and 50%: https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/294#issuecomment-2294454442

More details about the coloring were also added in the documentation: https://jso.dev/ADNLPModels.jl/dev/sparse/

The last release also uses the new major releases of SparseMatrixColorings.jl (v0.4.0) and SparseConnectivyTracer.jl (v0.6.1)
which lead more speed-ups and robustness.

**Reply by jbcaillau** [reply]
thanks @amontoison sounds timely and promising 👍🏽🤞🏽!

**Reply by ocots** [reply]
Thanks for the release and for the tuto!

**Reply by jbcaillau** [reply]
@PierreMartinon would be nice to re-run your previous tests with the last update 
https://github.com/control-toolbox/CTDirect.jl/discussions/111#discussioncomment-10102811

**Comment by jbcaillau** [comment]
https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/issues/165

**Comment by PierreMartinon** [comment]
NB. I updated the benchmark and removed 2 rather instable problems. 
Runs much faster now, so grid sizes were increased compared to the previous version.

| N     | 250 | 500 | 1000 | 2500 | 5000 | 10000 |
|-------|-----|-----|------|------|------|-------|
| 0.8.5 | 1.6 | 4.1 | 11.5 | 36.2 | 95.7 | 271.7 |
| 0.8.7 | 2.0 | 3.6 | 7.8  | 29.3 | 79.1 | 197.2 |
| speedup |0.8| 1.14| 1.47 | 1.24 | 1.21 | 1.38  |

Seems we got a 20-40% speed gain, nice !

**Reply by jbcaillau** [reply]
good, thanks @PierreMartinon 👍🏽

**Comment by amontoison** [comment]
New benchmarks on the ACOPF problems:
- https://discourse.julialang.org/t/ac-optimal-power-flow-in-various-nonlinear-optimization-frameworks/78486/98

I think that replacing `ReverseDiff.jl` with `Enzyme.jl` will be the easiest way to gain a speed-up in the future.
We have only updated sparsity detection and coloring since last year.

**Reply by gdalle** [reply]
I'm currently working on multiple arguments for DifferentiationInterface, which was the last hurdle for the hessian of the lagrangian with a fixed multiplicator. Once that lands, I think we can consider the switch to DI

**Reply by amontoison** [reply]
I will discuss about the Hessian of the Lagrangian with William Moses during AD2024.
Maybe we can optimize things at low-level.

**Reply by gdalle** [reply]
In that case such optimizations are likely to land either in Enzyme or in DI's Enzyme extension 😉

**Comment by PierreMartinon** [comment]
Allright, benchmark with the new CTModels (abstract) !

Trapeze
```
julia> bench()
Grid size    250: time (s) =    0.5
Grid size    500: time (s) =    1.4
Grid size   1000: time (s) =    3.5
Grid size   2500: time (s) =   17.3
Grid size   5000: time (s) =   71.0

julia> bench(;adnlp_backend=:manual)
Grid size    250: time (s) =    0.8
Grid size    500: time (s) =    1.7
Grid size   1000: time (s) =    3.5
Grid size   2500: time (s) =   11.3
Grid size   5000: time (s) =   24.0
```
Midpoint
```
julia> bench(;disc_method=:midpoint)
Grid size    250: time (s) =    0.7
Grid size    500: time (s) =    1.6
Grid size   1000: time (s) =    4.1
Grid size   2500: time (s) =   17.9
Grid size   5000: time (s) =   62.2

julia> bench(;disc_method=:midpoint, adnlp_backend=:manual)
Grid size    250: time (s) =    0.8
Grid size    500: time (s) =    1.9
Grid size   1000: time (s) =    3.8
Grid size   2500: time (s) =   11.6
Grid size   5000: time (s) =   24.2
```
Gauss Legendre 2
```
julia> bench(;disc_method=:gauss_legendre_2)
Grid size    250: time (s) =    2.7
Grid size    500: time (s) =    9.1
Grid size   1000: time (s) =   22.5
<grid size 2500 fails one of the tests -_->
Grid size   5000: time (s) =  366.6

julia> bench(;disc_method=:gauss_legendre_2, adnlp_backend=:manual)
Grid size    250: time (s) =    3.3
Grid size    500: time (s) =    8.0
Grid size   1000: time (s) =   18.5
Grid size   2500: time (s) =   55.5
Grid size   5000: time (s) =  131.0
```
Gauss Legendre 3
```
julia> bench(;disc_method=:gauss_legendre_3)
Grid size    250: time (s) =    5.1
Grid size    500: time (s) =   15.9
Grid size   1000: time (s) =   34.4
Grid size   2500: time (s) =  160.4
Grid size   5000: time (s) =  624.0

julia> bench(;disc_method=:gauss_legendre_3, adnlp_backend=:manual)
Grid size    250: time (s) =    6.3
Grid size    500: time (s) =   15.8
Grid size   1000: time (s) =   38.7
Grid size   2500: time (s) =  104.9
Grid size   5000: time (s) =  194.1
```

**Reply by jbcaillau** [reply]
super nice @PierreMartinon 👍🏽 extra wish 🎅🏾: 
- would like to see allocations 🙏🏽
- see that trapeze is always slightly faster than its dual midpoint scheme (with the advantage that trapeze has the same gridpoints)
- for the same optim tolerance, higher order schemes are slower but is there a gain at least in the number of iterations? (= would be nice to display time + alloc + iteration number 🥲)

**Comment by jbcaillau** [comment]
@rand-asswad please give a try to the new high order schemes on your problem 🤞🏽
